[RFC] Lint codebase using standard

tvararu opened the pull request:
Depends on:

- [x] https://github.com/alphagov/govuk_prototype_kit/pull/223
- [x] https://github.com/alphagov/govuk_prototype_kit/pull/233

This PR is an open discussion on the topic of using a JavaScript linting tool across our codebases.

I throw my hat in the ring by suggesting [standard](https://github.com/feross/standard) and demonstrating with this PR how an implementation would work for a node project.

Code was automatically converted using https://github.com/maxogden/standard-format.

We already have [govuk-lint](https://github.com/alphagov/govuk-lint) whose responsibility it is to handle linting concerns, and I'd love to hear what our thoughts are on relying on that instead.

nickcolley commented on app/assets/javascripts/details.polyfill.js:
```diff
@@ -6,188 +6,185 @@
 
-(function () {
-  'use strict';
+;(function () {
```
Shouldn't need to do this. https://github.com/twbs/bootstrap/issues/3057

paulmsmith commented:
I saw this tweet recently: https://twitter.com/brendaneich/status/741277499376603138 - Tim Berners-lee & Brendan Eich use standard.

joelanman commented:
thanks for this, sorry I dont have much experience on this - can you explain a bit more about how it works?

tvararu commented:
@joelanman certainly!

https://github.com/alphagov/govuk_prototype_kit/pull/224/commits/1edbe61ad6a3c22be101becc71d25095a04b8315 is where I've added `standard` as a project dependency. [standard](http://standardjs.com/) is an opinionated JavaScript linter that uses [eslint](http://eslint.org/) underneath and acts as a set of sensible defaults for it; [here are the most important rules](https://github.com/feross/standard#rules).

Since `standard` is based on `eslint`, overriding rules inside a file / ignoring files works through the same mechanisms, and if we want to move away or build on top of its opinionated ruleset, we can easily do so by swapping it with `eslint` and dropping an [.eslintrc file](https://github.com/feross/eslint-config-standard/blob/master/eslintrc.json).

Also worth mentioning some other "batteries-included" linters that fork standard or are in a similar vein:

- [semistandard](https://github.com/Flet/semistandard) (standard with semicolons)
- [happiness](https://github.com/JedWatson/happiness) (standard with semicolons and tabs)
- [doublestandard](https://github.com/Flet/doublestandard) (standard with 2 semicolons 💃)
- [uber-standard](https://www.npmjs.com/package/uber-standard) (standard forked by the Uber team)
- [eslint-config-airbnb](https://github.com/airbnb/javascript) (an eslint plugin based on Airbnb's styleguide)

I've hooked it up to `npm` scripts because I was hoping on `travis` to run `npm test` on our behalf on CI, but it looks like it relies on `grunt` to do it instead. I'll update it.

https://github.com/alphagov/govuk_prototype_kit/pull/224/commits/661776a7eae86912e7904533e43597ab0ee11a47 is the monster commit where I run [standard-format](https://github.com/maxogden/standard-format) on every file, and complete it with a few manual modifications that it's not smart enough to do by itself. Stuff I've had to do is:

- add `path.join()` calls where `__dirname + '/foo'` was being used
- split `var a,\nb,\nc` declarations into multiple ones
- add `/* global MyLib */` calls to make it clear that stuff like jQuery or GOVUK is being read from the global context

Hope I didn't waffle too much, I can update the commit/PR messages with more info if any of this is useful to have for posterity.

joelanman commented:
thanks! When does the linter run? Or do you have to run it manually?

tvararu commented:
@joelanman my hope is to update this so that it runs as part of CI, and combined with [text editor plugins](http://standardjs.com/index.html#text-editor-plugins) that's usually enough for projects I've used it in with other developers. Optionally we could run the test suite on a `pre-push` hook, to save a bit of time and get less red ❌'s on pull requests. (I can add that fairly easily if everyone likes it)

tvararu commented:
TODO:

- Hook `npm test` into CI
- Add documentation about manually running `npm run lint` or adding editor plugins

tvararu commented:
@joelanman I've written some documentation here: https://github.com/alphagov/govuk_prototype_kit/pull/224/commits/252b8b4d71733e2ed5018e9c2dc498ae8beb183e

This PR is good to go once https://github.com/alphagov/govuk_prototype_kit/pull/223 and https://github.com/alphagov/govuk_prototype_kit/pull/233 are merged.

nickcolley commented:
@tvararu like the addition in the docs, do you think we could briefly add why we're now linting and why standard being so opinionated is a good thing?

tvararu commented:
@nickcolley Updated the latest commit, added two more blurbs on "Why lint?" and "Why standard?"

robinwhittleton commented:
Spoke about this in #frontend. The specific problem this solves is standardising the JS that goes into govuk_prototype_kit, it’s not necessarily being suggested for actual prototypes (although of course people using the kit are free to adopt for their own code if they wish).

dsingleton commented:
Hugely pro some kind of standardised linting across our projects. 👍 

As discussed in the FE meeting, `standard` seems like it might be a good default. Happy to try it for a while on this repo, see how it feels, and roll it out to more repos if it works for us.

tvararu commented:
Thanks for the support @dsingleton!

I recall @edwardhorsford mentioning that he'd prefer not to link to the Linting doc from the main page of the README, because it's too technical. There are also a few other things I recall:

- [ ] Update README, remove link to "Linting", link from the CONTRIBUTING file instead
- [ ] Update Linting.md, expand on "Why standard" section (I don't remember exactly what though, @dsingleton to provide guidance)
- [ ] Update Linting.md, provide clear instructions for how to add linting to two popular editors (Sublime Text and Atom) and add a link to the official docs for other editors

Anybody subscribed to this thread feel free to edit this comment with things I've forgotten.

gemmaleigh commented:
If you'd like another project to try this out on, you are more than welcome to do the same for [GOV.UK elements](https://github.com/alphagov/govuk_elements)!

Aside from the JS from the govuk_frontend_toolkit, some of the govuk_prototype_kit JS has been directly copied from govuk_elements. 

If we have JS that can be of benefit to more than one project, it makes sense to move it to govuk_frontend_toolkit. 

An example of this is the [show/hide js](https://github.com/alphagov/govuk_frontend_toolkit/pull/305), which will replace [what is exists  in application.js](https://github.com/alphagov/govuk_prototype_kit/blob/master/app/assets/javascripts/application.js#L121).

dsingleton commented:
> Update Linting.md, expand on "Why standard" section (I don't remember exactly what though, @dsingleton to provide guidance)

My train of thought around this was: Linting rules can be a contentious subject, and a personal preference. We should document why we chose a particular set of rules over others - why `standard` was a good fit for us (rather than why it's good in general). If we make sure people can understand the reasons behind that choice we're less likely to have rehash discussions over the choice.

Hopefully that helps?

joelanman commented:
Happy to merge this if others are?

tvararu commented:
@joelanman yes please! ✨

joelanman commented:
@tvararu can you rebase against master and I'll merge

joelanman commented:
thanks everyone! 💃
